### PR TITLE
added `compresion.level' configuration option for fine-tuning gzip and LZ4

### PR DESCRIPTION
**Pull request by erkoln**
Hello,

librdkafka supports message sets compressed with either gzip, LZ4 or Snappy but currently provides no control over compression level used by the underlying algorithm when producing messages, meaning the default compression level is always used.

Default zlib level provides a good compression ratio, but does so at the cost of high CPU usage. When producing messages to a remote cluster, CPU can easily become a bottleneck for per-partition throughput.
On the other hand, Snappy and LZ4 use very little CPU but the compression ratio is not so good when compared to gzip. In this case, network bandwidth quickly becomes the bottleneck.

This patch allows compression level to be fine-tuned for LZ4 and gzip (this seemingly can't be controlled in the Snappy implementation used by librdkafka), thus making it possible to trade between CPU and bandwidth usage.

In my use-case, given the particular CPU and network constraints I have to do with, level-1 gzip typically provide the best overall message throughput, while providing about the same compression ratio as default-level gzip.

### CODE REVIEW

File: src/rdkafka_conf.c [hunk 1]
```diff
@@ -779,6 +779,25 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
 		} },
         { _RK_GLOBAL|_RK_PRODUCER, "compression.type", _RK_C_ALIAS,
           .sdef = "compression.codec" },
+	{ _RK_GLOBAL|_RK_PRODUCER, "compression.level", _RK_C_INT,
```
Comment (edenhill) [inline src/rdkafka_conf.c]: librdkafka now supports default topic configuration to be set on the global object (it instanties a default topic config object itself), we don't need to add this to the global config but only to the topic one.

File: src/rdkafka_conf.c [hunk 2]
```diff
@@ -779,6 +779,25 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
...
+	  _RK(compression_level),
+	  "compression level parameter for algorithm selected by configuration "
```
Comment (edenhill) [inline src/rdkafka_conf.c]: "Compression .."

File: src/rdkafka_lz4.c [hunk 3]
```diff
@@ -325,7 +325,7 @@ rd_kafka_lz4_decompress (rd_kafka_broker_t *rkb, int proper_hc, int64_t Offset,
  * @returns allocated buffer in \p *outbuf, length in \p *outlenp.
  */
 rd_kafka_resp_err_t
-rd_kafka_lz4_compress (rd_kafka_broker_t *rkb, int proper_hc,
+rd_kafka_lz4_compress (rd_kafka_broker_t *rkb, int proper_hc, int compLevel,
```
Comment (edenhill) [inline src/rdkafka_lz4.c]: no camel casing, use snake casing comp_level

File: src/rdkafka_msgset_writer.c [hunk 4]
```diff
@@ -794,9 +794,14 @@ rd_kafka_msgset_writer_compress_gzip (rd_kafka_msgset_writer_t *msetw,
         const void *p;
         size_t rlen;
         int r;
+        int compLevel =
+                msetw->msetw_rktp->rktp_rkt->rkt_conf.compression_level;
+        if (compLevel>Z_BEST_COMPRESSION) {
```
Comment (edenhill) [inline src/rdkafka_msgset_writer.c]: spaces: `if (comp_level > Z_BEST_COMPRESSION)`

File: src/rdkafka_msgset_writer.c [hunk 5]
```diff
@@ -931,9 +936,18 @@ static int
 rd_kafka_msgset_writer_compress_lz4 (rd_kafka_msgset_writer_t *msetw,
                                      rd_slice_t *slice, struct iovec *ciov) {
         rd_kafka_resp_err_t err;
+        int compLevel =
+                msetw->msetw_rktp->rktp_rkt->rkt_conf.compression_level;
+        if (compLevel>RD_KAFKA_COMPLEVEL_LZ4_MAX) {
+                compLevel = RD_KAFKA_COMPLEVEL_LZ4_MAX;
+        }
```
Comment (edenhill) [inline src/rdkafka_msgset_writer.c]: `} else {`

Please see the style guide: https://github.com/edenhill/librdkafka/blob/master/CONTRIBUTING.md#follow-code-style

File: src/rdkafka_msgset_writer.c [hunk 6]
```diff
@@ -931,9 +936,18 @@ static int
 rd_kafka_msgset_writer_compress_lz4 (rd_kafka_msgset_writer_t *msetw,
                                      rd_slice_t *slice, struct iovec *ciov) {
         rd_kafka_resp_err_t err;
+        int compLevel =
+                msetw->msetw_rktp->rktp_rkt->rkt_conf.compression_level;
+        if (compLevel>RD_KAFKA_COMPLEVEL_LZ4_MAX) {
```
Comment (edenhill) [inline src/rdkafka_msgset_writer.c]: Normalize these values in topic_new() instead so it doesn't need to be done for each messageset

File: src/rdkafka_conf.c (hunk 1 above)
Comment (erkoln) [reply src/rdkafka_conf.c]: OK, done in 81b010d
What about `compression.codec`. Shouldn't it be the same?

File: src/rdkafka_conf.c (hunk 2 above)
Comment (erkoln) [reply src/rdkafka_conf.c]: 5523bda :)

File: src/rdkafka_msgset_writer.c (hunk 6 above)
Comment (erkoln) [reply src/rdkafka_msgset_writer.c]: Right! done in f15ebaa

File: src/rdkafka_lz4.c (hunk 3 above)
Comment (erkoln) [reply src/rdkafka_lz4.c]: a73440a

File: src/rdkafka_msgset_writer.c (hunk 4 above)
Comment (erkoln) [reply src/rdkafka_msgset_writer.c]: a73440a

File: src/rdkafka_msgset_writer.c (hunk 5 above)
Comment (erkoln) [reply src/rdkafka_msgset_writer.c]: a73440a

**Comment by erkoln** [comment]
Hello,
I rebased my fork on current master to avoid polluting the commit log in case of merge.
This seems to have broken change requests tracking but I tried to reply with relevant commit hashes.
Sorry for that

File: src/rdkafka_conf.c (hunk 1 above)
Comment (erkoln) [reply src/rdkafka_conf.c]: Oups! Forgot to actually remove the field from config structure. Done in f7584f8

File: src/rdkafka_lz4.c [hunk 7]
```diff
@@ -339,8 +339,11 @@ rd_kafka_lz4_compress (rd_kafka_broker_t *rkb, int proper_hc,
 
         /* Required by Kafka */
         const LZ4F_preferences_t prefs =
-                { .frameInfo = { .blockMode = LZ4F_blockIndependent } };
-
+                {
+												.frameInfo = { .blockMode = LZ4F_blockIndependent },
```
Comment (edenhill) [inline src/rdkafka_lz4.c]: indent looks weird here

File: src/rdkafka_topic.c [hunk 8]
```diff
@@ -313,6 +313,25 @@ shptr_rd_kafka_itopic_t *rd_kafka_topic_new0 (rd_kafka_t *rk,
 	if (rkt->rkt_conf.compression_codec == RD_KAFKA_COMPRESSION_INHERIT)
 		rkt->rkt_conf.compression_codec = rk->rk_conf.compression_codec;
 
+				switch (rkt->rkt_conf.compression_codec) {
+				case RD_KAFKA_COMPRESSION_GZIP:
+								if(rkt->rkt_conf.compression_level > RD_KAFKA_COMPLEVEL_GZIP_MAX)
```
Comment (edenhill) [inline src/rdkafka_topic.c]: indentation is off here.
Also check style guide, space between if and paren

File: src/rdkafka_topic.c [hunk 9]
```diff
@@ -313,6 +313,25 @@ shptr_rd_kafka_itopic_t *rd_kafka_topic_new0 (rd_kafka_t *rk,
...
+												rkt->rkt_conf.compression_level =
+																RD_KAFKA_COMPLEVEL_GZIP_MAX;
+								break;
```
Comment (edenhill) [inline src/rdkafka_topic.c]: There doesnt seem to be any code to set the default compression level for gzip

File: src/rdkafka_conf.c [hunk 10]
```diff
@@ -923,6 +923,23 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
 		} },
         { _RK_TOPIC | _RK_PRODUCER, "compression.type", _RK_C_ALIAS,
           .sdef = "compression.codec" },
+	{ _RK_TOPIC|_RK_PRODUCER, "compression.level", _RK_C_INT,
+	  _RKT(compression_level),
+	  "Compression level parameter for algorithm selected by configuration "
+	  "property `compression.codec`. Higher values will result in better "
+	  "compression at the cost of more CPU usage. Usable range is "
+	  "algorithm-dependent: "
+#if WITH_ZLIB
```
Comment (edenhill) [inline src/rdkafka_conf.c]: You can remove the ifdefs here, I think it is fine to provide these ranges in the config doc even if corresponding support isn't included.

**Comment by erkoln** [comment]
Squashed history, which will hopefully also fix PR status.

**Comment by edenhill** [comment]
@andoma What's up with the mips builder?

File: src/rdkafka_topic.c [hunk 11]
```diff
@@ -313,6 +313,30 @@ shptr_rd_kafka_itopic_t *rd_kafka_topic_new0 (rd_kafka_t *rk,
 	if (rkt->rkt_conf.compression_codec == RD_KAFKA_COMPRESSION_INHERIT)
 		rkt->rkt_conf.compression_codec = rk->rk_conf.compression_codec;
 
+        switch (rkt->rkt_conf.compression_codec) {
+        case RD_KAFKA_COMPRESSION_GZIP:
+                /* RD_KAFKA_COMPLEVEL_DEFAULT == Z_DEFAULT_COMPRESSION so there
```
Comment (edenhill) [inline src/rdkafka_topic.c]: That's leaking abstraction, we should not rely on the value of Z_DEFAULT_COMPRESSION being equal our value, unless we explicitly it define it to be the same, which we don't (and shouldn't).

So I'd rather see us handling the default case explicitly here

File: src/rdkafka_topic.c [hunk 12]
```diff
@@ -313,6 +313,30 @@ shptr_rd_kafka_itopic_t *rd_kafka_topic_new0 (rd_kafka_t *rk,
...
+                        rkt->rkt_conf.compression_level =
+                                RD_KAFKA_COMPLEVEL_GZIP_MAX;
+                break;
+        case RD_KAFKA_COMPRESSION_LZ4:
+                /* For LZ4, default compression level is 0 and needs mapping:
+                 * check upper and lower bounds. */
+                if (rkt->rkt_conf.compression_level > RD_KAFKA_COMPLEVEL_LZ4_MAX)
+                        rkt->rkt_conf.compression_level =
+                                RD_KAFKA_COMPLEVEL_LZ4_MAX;
+                else if (rkt->rkt_conf.compression_level < 0)
```
Comment (edenhill) [inline src/rdkafka_topic.c]: `== RD_KAFKA_COMPLEVEL_DEFAULT`

File: src/rdkafka_topic.c [hunk 13]
```diff
@@ -313,6 +313,30 @@ shptr_rd_kafka_itopic_t *rd_kafka_topic_new0 (rd_kafka_t *rk,
...
+                        rkt->rkt_conf.compression_level = 0;
+                break;
+        case RD_KAFKA_COMPRESSION_SNAPPY:
+        default:
+                /* Compression level has no effet in this case */
```
Comment (edenhill) [inline src/rdkafka_topic.c]: effect

File: src/rdkafka_topic.c (hunk 11 above)
Comment (erkoln) [reply src/rdkafka_topic.c]: done, let me know if that's better :)

File: src/rdkafka_topic.c (hunk 12 above)
Comment (erkoln) [reply src/rdkafka_topic.c]: okay, done (5c90922)

File: src/rdkafka_topic.c (hunk 13 above)
Comment (erkoln) [reply src/rdkafka_topic.c]: fixed (5c90922)
oh the joy of coding while being sleep deprived...

**Comment by edenhill** [comment]
Thank you for this PR!

**Comment by erkoln** [comment]
You're welcome. :-)
Thank you for the great piece of software that librdkafka is!

**Pull request merged**